src/uu/cp/src/cp.rs

```diff
@@ -1666,17 +1651,40 @@ fn copy_on_write_macos(
 }
 
 /// Generate an error message if `target` is not the correct `target_type`
-pub fn verify_target_type(target: &Path, target_type: &TargetType) -> CopyResult<()> {
-    match (target_type, target.is_dir()) {
-        (&TargetType::Directory, false) => {
-            Err(format!("target: {} is not a directory", target.quote()).into())
+pub fn verify_target_type(sources: &[Source], target: &TargetSlice) -> CopyResult<TargetType> {
+    let target_path_string = target.to_string_lossy();
+    let seems_like_dir = target_path_string.ends_with('/')
+        || target_path_string == "."
+        || target_path_string == ".."
+        || sources.len() > 1;
+    match (
+        seems_like_dir,
+        target.exists(),
+        target.is_dir(),
+        // replace with target.is_symlink() in rust >= 1.58
+        if let Ok(metadata) = fs::symlink_metadata(target) {
+            metadata.file_type().is_symlink()
+        } else {
+            false
+        }, // target.is_symlink(),
+    ) {
+        (true, false, ..) => {
+            return Err(format!("directory {} does not exist", target.quote()).into())
         }
-        (&TargetType::File, true) => Err(format!(
-            "cannot overwrite directory {} with non-directory",
-            target.quote()
-        )
-        .into()),
-        _ => Ok(()),
+        (true, true, false, false) => {
+            return Err(format!("target: {} is not a directory", target.quote()).into())
+        }
+        (true, true, false, true) => {
+            return Err(format!(
+                "destination {} is a symlink to a regular file",
+                target.quote()
+            )
+            .into())
+        }
+        (true, true, true, _) => Ok(TargetType::Directory),
+        (false, _, false, _) => Ok(TargetType::File),
+        (false, true, true, _) => Ok(TargetType::Directory),
+        _ => Err("cannot determine target type".into()),
```

Comment:
The missing cases seem to be `(false, false, true, _)`. I.e. does not exist but is a dir, which is obviously not possible. This is kind of a downside of putting all booleans in the match. I think it'd be good to write this case out explicitly, explain in a comment why it's impossible and then put a `unreachable!` here instead of an `Err`.